### PR TITLE
Develop - SFTPHandlerReplacePlugin feature

### PR DESCRIPTION
**Pull request by The5imon**
Can replace SFTP transmitted files when communicating over the ssh proxy server.

Following cases are known problems:
- PUT where replacement file size > requested file (remote file is closed before all data can be written)

### CODE REVIEW

File: ssh_proxy_server/cli.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 
 from paramiko import Transport
 
+from ssh_proxy_server.clients.ssh import SSHClient, BaseSSHClient
```
Comment (manfred-kaiser) [inline ssh_proxy_server/cli.py:6]: Import von "SSHClient" und "BaseSSHClient"  wird nicht benötigt. Kann somit entfernt werden

File: ssh_proxy_server/forwarders/sftp.py [hunk 2]
```diff
@@ -67,3 +70,42 @@ def write(self, offset, data):
...
+        logging.info("intercepting sftp file, replacement: %s", self.args.sftp_replacement_file)
+        self.file = None
+        self.replacement = open(self.args.sftp_replacement_file, "rb")
+        self.buf_size = None
+
+    def close(self):
+        self.replacement.close()
+
+    def handle_data(self, data):
+        # Cannot reasonably detect given buffer size, using best guess: biggest power of 2
```
Comment (manfred-kaiser) [inline ssh_proxy_server/forwarders/sftp.py:100]: Why are you using this workarround with biggest power of 2?

This method is thread safe, and your file descriptor should remeber its current position. So you can read the same amount of data as given from the remote client/server

**Review by manfred-kaiser — CHANGES_REQUESTED** [review]

File: ssh_proxy_server/forwarders/sftp.py (hunk 2 above)
Comment (The5imon) [reply ssh_proxy_server/forwarders/sftp.py:100]: Yes, in theory, that was the first thing I tried and tested, but for some reason the SFTP Client on the host will reject and close the SFTP channel if it receives more than an arbitrary amount of data.
An error that occurs mainly with small files. For example, when passing through data from the replacement file with a buffer size as big as the amount of data that was requested at the initial transferring of the file, following error message will be thrown:

`Received more data than asked for 141 > 56`  

Keep in mind that the file has a size of just 141 bytes.

Comment (manfred-kaiser) [reply ssh_proxy_server/forwarders/sftp.py]: small files and files up to the size of the wanted files are transfered without errors. Only larger files does not work.
Tested with commit: 598e2d22eb6c4543a4942dd6992d359b4a4ab993

File: ssh_proxy_server/forwarders/sftp.py [hunk 3]
```diff
@@ -67,3 +69,43 @@ def write(self, offset, data):
...
+        logging.info("sftp file transfer detected: %s", filename)
+        logging.info("intercepting sftp file, replacement: %s", self.args.sftp_replacement_file)
+        self.file = None
+        self.replacement = open(self.args.sftp_replacement_file, "rb")
+        self.buf_size = None
+
+    def close(self):
+        self.replacement.close()
+
+    def handle_data(self, data):
```
Comment (manfred-kaiser) [inline ssh_proxy_server/forwarders/sftp.py:98]: you have to pass the data length to the method:

```python
    def handle_data(self, data, length=None):
        return self.replacement.read(length)
```
To pass the legth parameter, you have to modify the 'handle_data' method from 'SFTPBaseHandle':

```python
class SFTPBaseHandle(paramiko.SFTPHandle):

    def read(self, offset, length):
        data = self.readfile.read(length)
        return self.plugin.handle_data(data, length)
```

To get the file replacement file size, you have to subclass  'ssh_proxy_server.interfaces' and edit the methods 'stat' and 'lstat':

```python
from paramiko.sftp_attr import SFTPAttributes

class ReplaceerSFTPProxyServerInterface(SFTPProxyServerInterface):

    def lstat(self, path):
        self.session.sftp_client_ready.wait()
        stat_remote = self.session.sftp_client.lstat(path)
        stat_replace = SFTPAttributes.from_stat(os.stat('/bin/bash'))
        stat_remote.st_size = stat_replace.st_size
        return stat_remote

    def stat(self, remotePath):
        self.session.sftp_client_ready.wait()
        stat_remote = self.session.sftp_client.stat(remotePath)
        stat_replace = SFTPAttributes.from_stat(os.stat('/bin/bash'))
        stat_remote.st_size = stat_replace.st_size
        return stat_remote
```

You should also add a new cmd argument to the main-Function in cli.py, where you can specify the ProxyServerInterface.

**Review by manfred-kaiser — CHANGES_REQUESTED** [review]

File: ssh_proxy_server/forwarders/sftp.py (hunk 3 above)
Comment (manfred-kaiser) [reply ssh_proxy_server/forwarders/sftp.py:98]: http://docs.paramiko.org/en/stable/api/sftp.html

**Review by manfred-kaiser — APPROVED** [review]

**Pull request merged**